cdap-cli/src/main/java/co/cask/cdap/cli/command/AbstractSetPreferencesCommand.java

```diff
@@ -29,48 +30,93 @@
-        printSuccessMessage(printStream, type);
-        break;
-
-      default:
-        throw new IllegalArgumentException("Unrecognized Element Type for Preferences " + type.getTitleName());
+    private final PreferencesClient client;
+    private final ElementType type;
+
+    protected AbstractSetPreferencesCommand(ElementType type, PreferencesClient client, CLIConfig cliConfig) {
+        super(cliConfig);
+        this.type = type;
+        this.client = client;
+    }
+
+    protected abstract void printSuccessMessage(PrintStream printStream, ElementType type);
+
+    protected void setPreferences(String[] programIdParts, PrintStream printStream,
+                                  Map<String, String> args) throws Exception {
+        switch (type) {
+            case INSTANCE:
+                checkInputLength(programIdParts, 0);
+                client.setInstancePreferences(args);
+                printSuccessMessage(printStream, type);
+                break;
+
+            case NAMESPACE:
+                checkInputLength(programIdParts, 0);
+                client.setNamespacePreferences(cliConfig.getCurrentNamespace(), args);
+                printSuccessMessage(printStream, type);
+                break;
+
+            case APP:
+                client.setApplicationPreferences(parseAppId(programIdParts), args);
+                printSuccessMessage(printStream, type);
+                break;
+
+            case FLOW:
+            case MAPREDUCE:
+            case WORKFLOW:
+            case SERVICE:
+            case SPARK:
+                client.setProgramPreferences(parseProgramId(programIdParts, type.getProgramType()), args);
+                printSuccessMessage(printStream, type);
+                break;
+
+            default:
+                throw new IllegalArgumentException("Unrecognized Element Type for Preferences " + type.getTitleName());
+        }
+    }
+
+
+
+    protected String determinePattern(String action) { // action can either be set or load
+        if (action.equals("set")) {
+            switch (type) {
+                case INSTANCE:
+                    return String.format("set preferences %s <%s>", type.getName(), ArgumentName.RUNTIME_ARGS);
+                case NAMESPACE:
+                    return String.format("set preferences %s <%s>", type.getName(), ArgumentName.RUNTIME_ARGS);
+                case APP:
+                    return String.format("set preferences %s <%s> <%s>", type.getName(), ArgumentName.RUNTIME_ARGS,
+                            type.getArgumentName());
+                case FLOW:
+                case MAPREDUCE:
+                case WORKFLOW:
+                case SERVICE:
+                case SPARK:
+                    return String.format("set preferences %s <%s> [<%s>]", type.getName(), ArgumentName.RUNTIME_ARGS,
+                            type.getArgumentName());
+            }
```

Comment:
reading this, I'd rather see one return statement, and the first cases set extraParameters = "", the next extraParameters = "<%s>" and the last extraParameters = "[<%s>]"...